renku_notebooks/api/classes/data_service.py

```diff
@@ -246,3 +248,77 @@ def get_default_class(self) -> Dict[str, Any]:
 
     def find_acceptable_class(self, *args, **kwargs) -> Optional[ServerOptions]:
         return self.options
+
+
+@dataclass
+class GitProviderHelper:
+    """Calls to the data service to configure git providers."""
+
+    service_url: str
+    renku_url: str
+    internal_gitlab_url: str
+
+    def __post_init__(self):
+        self.service_url = self.service_url.rstrip("/")
+        self.renku_url = self.renku_url.rstrip("/")
+
+    def get_providers(self, user: User) -> list[GitProvider]:
+        connections = self.get_oauth2_connections(user=user)
+        providers: dict[str, GitProvider] = dict()
+        for c in connections:
+            if c.provider_id in providers:
+                continue
+            provider = self.get_oauth2_provider(c.provider_id)
+            access_token_url = urljoin(
+                self.renku_url,
+                urlparse(self.service_url + f"/oauth2/connections/{c.id}/token").path,
+            )
+            providers[c.provider_id] = GitProvider(
+                id=c.provider_id,
+                url=provider.url,
+                connection_id=c.id,
+                access_token_url=access_token_url,
+            )
+
+        providers_list = [p for p in providers.values()]
+        # Insert the internal GitLab as the first provider
+        internal_gitlab_access_token_url = urljoin(self.renku_url, "/api/auth/gitlab/exchange")
+        providers_list[:0] = [
+            GitProvider(
+                id=INTERNAL_GITLAB_PROVIDER,
+                url=self.internal_gitlab_url,
+                connection_id="",
+                access_token_url=internal_gitlab_access_token_url,
+            )
+        ]
+        return providers_list
+
+    def get_oauth2_connections(self, user: User | None = None) -> list[OAuth2Connection]:
+        if user is None or user.access_token is None:
+            return []
+        request_url = self.service_url + "/oauth2/connections"
+        headers = {"Authorization": f"bearer {user.access_token}"}
+        res = requests.get(request_url, headers=headers)
+        if res.status_code != 200:
+            raise IntermittentError(
+                message="The data service sent an unexpected response, please try again later"
+            )
+        connections = res.json()
+        connections = [OAuth2Connection.from_dict(c) for c in connections]
+        return [c for c in connections if c.status == "connected"]
+
+    def get_oauth2_provider(self, provider_id: str) -> OAuth2Provider:
+        request_url = self.service_url + f"/oauth2/providers/{provider_id}"
```

Comment:
move into f-string

Reply:
Done in d29ed41420c6b89d0ccbb60911289e2659412cd6.